CONTRIBUTING.md

```diff
@@ -0,0 +1,160 @@
+- Any tests creating resource should always cleanup via `AfterAll` or `AfterEach` blocks
+- Each test should have two primary Describe blocks and be tagged accordingly: UnitTests and IntegrationTests
+  - `Describe "$CommandName Unit Tests" -Tag 'UnitTests' {...}`
+  - `Describe "$commandname Integration Tests" -Tag "IntegrationTests" {...}`
+
+Tests make sure a "contract" is made between the code and its behavior: once a test is formalized, changes to the code itself or enhancement will be written making sure existing functionality is retained, making the entire dbatools experience more stable.
+
+You can inspect/copy/cannibalize existing tests. You'll see that every test file is named with a simple convention `Verb-Noun*.Tests.ps1`, and this is required by [Pester](https://GitHub.com/pester/Pester), which is the de-facto standard for running tests in PowerShell.
+
+## AppVeyor Environment
+
+AppVeyor is hooked up to test any commit, including PRs. Each commit triggers 5 builds, each referred to as a "scenario". We have the scenarios setup where the dbatools log is published as an artifact should you need to view why test are failing.
+
+- 2008R2 : a server with a single SQL Server 2008 R2 Express Edition instance available ($script:instance1)
+- 2016 : a server with a single SQL Server 2016 Developer Edition instance available ($script:instance2)
+- 2016_service: used to test service restarts
+- 2016_2017 : a server with two instances available, 2016 and 2017 Developer Edition
+- default: a server with two instances available, one SQL Server 2008 R2 Express Edition and a SQL Server 2016 Developer Edition
+
+Builds are split among "scenario"(s) because not every test requires everything to be up and running, and resources on AppVeyor are constrained.
+
+Ideally:
+
+1. Whenever possible, write UnitTests.
+1. You should write IntegrationTests ideally running in **EITHER** the 2008R2 or the 2016 "scenario".
+1. Default and 216_2017 are the most resource constrained and are left to run the Copy-* commands which are the only ones **needing** two active instances.
+1. If you want to write tests that, e.g, target **BOTH** 2008R2 and 2016, try to avoid writing tests that need both instances to be active at the same time.
+
+AppVeyor is set up to recognize what "scenario" is required by your test, simply inspecting for the presence of combinations of `$script:instance1`, `$script:instance2` and `$script:instance3`. If you need to fall into case (4), write two test files, e.g. _Get-DbaFoo.first.Tests.ps1_ (targeting `$script:instance1` only) and _Get-DbaFoo.second.Tests.ps1_ (targeting `$script:instance2` only).
+
+If you don't want to wait for the entire test suite to run (i.e. you need to run only `Get-DbaFoo` on AppVeyor), you can use a **_magic command_** within the commit message, namely `(do Get-DbaFoo)` . This will run only test files within the test folder matching this mask `tests\*Get-DbaFoo*.Tests.ps1`.
```

Comment:
I didn't know that! Very cool

Reply:
You can do this for your branch as well, includes forks.